src/celeritas/global/TrackLauncher.hh

```diff
@@ -16,48 +21,193 @@ namespace celeritas
+    //! \name Type aliases
+    using ParamsRef = NativeCRef<CoreParamsData>;
+    using StateRef = NativeRef<CoreStateData>;
+    //!@}
+
+  public:
+    CELER_FUNCTION
+    TrackLauncher(ParamsRef const& params, StateRef const& state, Ts... args)
+        : params_{params}
+        , state_{state}
+        , launch_impl_{celeritas::forward<Ts>(args)...}
+    {
+    }
+
+    CELER_FUNCTION void operator()(ThreadId thread) const
+    {
+        CELER_EXPECT(thread);
+#if CELER_DEVICE_COMPILE
+        if (!(thread < state_.size()))
+        {
+            return;
+        }
+#else
+        CELER_EXPECT(thread < state_.size());
+#endif
+        CoreTrackView const track(params_, state_, thread);
+
+        return launch_impl_(track);
+    }
+
+  private:
+    ParamsRef const& params_;
+    StateRef const& state_;
+    detail::TrackLauncherImpl<Ts...> launch_impl_;
+};
+
+//---------------------------------------------------------------------------//
+/*!
+ * Launch the track only when a certain condition applies.
+ *
+ * The condition C must have the signature \code
+ * <bool(SimTrackView const&)>
+  \endcode
+ *
+ * see \c make_active_track_launcher for an example where this is used to apply
+ * only to active (or killed) tracks.
+ */
+template<class C, class... Ts>
+class ConditionalTrackLauncher
+{
+  public:
+    //!@{
+    //! \name Type aliases
+    using ParamsRef = NativeCRef<CoreParamsData>;
+    using StateRef = NativeRef<CoreStateData>;
+    //!@}
+
+  public:
+    CELER_FUNCTION
+    ConditionalTrackLauncher(ParamsRef const& params,
+                             StateRef const& state,
+                             C applies,
+                             Ts... args)
+        : params_{params}
+        , state_{state}
+        , applies_{celeritas::forward<C>(applies)}
+        , launch_impl_{celeritas::forward<Ts>(args)...}
+    {
+    }
+
+    CELER_FUNCTION void operator()(ThreadId thread)
+    {
+        CELER_EXPECT(thread);
+#if CELER_DEVICE_COMPILE
+        if (!(thread < state_.size()))
+        {
+            return;
+        }
+#else
+        CELER_EXPECT(thread < state_.size());
+#endif
+        CoreTrackView const track(params_, state_, thread);
+        if (!applies_(track.make_sim_view()))
+        {
+            return;
+        }
+
+        return launch_impl_(track);
+    }
+
+  private:
+    ParamsRef const& params_;
+    StateRef const& state_;
+    C applies_;
+    detail::TrackLauncherImpl<Ts...> launch_impl_;
+};
+
+//---------------------------------------------------------------------------//
+// DEDUCTION GUIDES
+//---------------------------------------------------------------------------//
+template<class... Ts>
+CELER_FUNCTION TrackLauncher(NativeCRef<CoreParamsData> const&,
+                             NativeRef<CoreStateData> const&,
+                             Ts...)
+    ->TrackLauncher<Ts...>;
+
+template<class C, class... Ts>
+CELER_FUNCTION ConditionalTrackLauncher(NativeCRef<CoreParamsData> const&,
```

Comment:
Interesting that you have to use `__host__ __device__` annotations for HIP since deduction guides are not functions according to the standard.

Reply:
Yeah that surprised me too...